cpp/tests/copying/reverse_tests.cpp

```diff
@@ -0,0 +1,175 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_utilities.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/cudf_gtest.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/copying.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+template <typename T>
+class ReverseTypedTestFixture : public cudf::test::BaseFixture {
+};
+
+TYPED_TEST_CASE(ReverseTypedTestFixture, cudf::test::AllTypes);
+TYPED_TEST(ReverseTypedTestFixture, ReverseTable)
+{
+  using T = TypeParam;
+  constexpr cudf::size_type num_values{10};
+
+  auto input = cudf::test::fixed_width_column_wrapper<T, int32_t>(
+    thrust::make_counting_iterator(0), thrust::make_counting_iterator(0) + num_values);
+
+  auto expected_elements = cudf::detail::make_counting_transform_iterator(
+    0, [num_values] __device__(auto i) { return num_values - i - 1; });
+
+  auto expected =
+    cudf::test::fixed_width_column_wrapper<T, typename decltype(expected_elements)::value_type>(
+      expected_elements, expected_elements + num_values);
+
+  auto input_table = cudf::table_view{{input}};
+  auto const p_ret = cudf::reverse(input_table);
+
+  EXPECT_EQ(p_ret->num_columns(), 1);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(p_ret->view().column(0), expected, true);
+}
+
+TYPED_TEST(ReverseTypedTestFixture, ReverseColumn)
+{
+  using T = TypeParam;
+  constexpr cudf::size_type num_values{10};
+
+  auto input = cudf::test::fixed_width_column_wrapper<T, int32_t>(
+    thrust::make_counting_iterator(0), thrust::make_counting_iterator(0) + num_values);
+
+  auto expected_elements = cudf::detail::make_counting_transform_iterator(
+    0, [num_values] __device__(auto i) { return num_values - i - 1; });
+
+  auto expected =
+    cudf::test::fixed_width_column_wrapper<T, typename decltype(expected_elements)::value_type>(
+      expected_elements, expected_elements + num_values);
+
+  auto const column_ret = cudf::reverse(input);
+
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(column_ret->view(), expected, true);
+}
+
+TYPED_TEST(ReverseTypedTestFixture, ReverseNullable)
+{
+  using T = TypeParam;
+  constexpr cudf::size_type num_values{20};
+
+  std::vector<int64_t> input_values(num_values);
+  std::iota(input_values.begin(), input_values.end(), 1);
+
+  std::vector<bool> input_valids(num_values);
+  for (size_t i{0}; i < input_valids.size(); i++) { input_valids[i] = (i % 2) == 0 ? true : false; }
+
+  std::vector<T> expected_values;
+  std::vector<bool> expected_valids;
+  for (int i = num_values - 1; i > -1; i--) {
+    expected_values.push_back(cudf::test::make_type_param_scalar<T>(input_values[i]));
+    expected_valids.push_back(input_valids[i]);
+  }
```

Comment:
See, named algorithms make your intent very clear here, and prevent future errors modifying tricky backward indexing:

```suggestion
  std::vector<T> expected_values(input_values.size());
  thrust::host_vector<bool> expected_valids(input_valids.size());
  std::copy_backward(input_values.begin(), input_values.end(), expected_values.begin());
  std::copy_backward(input_valids.begin(), input_valids.end(), expected_valids.begin());
```

Reply:
Gotcha! This does look much cleaner.